drivers/counter/counter_nxp_pit.c

```diff
@@ -0,0 +1,407 @@
+
+	if (config->base->CHANNEL[channel].TCTRL & PIT_TCTRL_TEN_MASK) {
+		/* Timer already enabled, check flags before resetting */
+		if (cfg->flags & COUNTER_TOP_CFG_DONT_RESET) {
+			return -ENOTSUP;
+		}
+		PIT_StopTimer(config->base, channel);
+		PIT_SetTimerPeriod(config->base, channel, cfg->ticks);
+		PIT_StartTimer(config->base, channel);
+	} else {
+		PIT_SetTimerPeriod(config->base, channel, cfg->ticks);
+	}
+
+	return 0;
+}
+
+static uint32_t nxp_pit_get_pending_int(const struct device *dev)
+{
+	const struct nxp_pit_config *config = dev->config;
+	uint32_t mask = PIT_TFLG_TIF_MASK;
+	uint32_t flags;
+	int channel_id = PIT_CHANNEL_ID(dev);
+
+	flags = PIT_GetStatusFlags(config->base, channel_id);
+
+	return ((flags & mask) == mask);
+}
+
+static uint32_t nxp_pit_get_frequency(const struct device *dev)
+{
+	const struct nxp_pit_config *config = dev->config;
+	uint32_t clock_rate;
+
+	if (clock_control_get_rate(config->clock_dev, config->clock_subsys, &clock_rate)) {
+		LOG_ERR("Failed to get clock rate");
+		return 0;
+	}
+
+	return clock_rate;
+}
+
+#if DT_NODE_HAS_PROP(DT_COMPAT_GET_ANY_STATUS_OKAY(nxp_pit), interrupts)
+static void nxp_pit_isr(const struct device *dev)
+{
+	const struct nxp_pit_config *config = dev->config;
+	uint32_t flags;
+
+	LOG_DBG("pit counter isr");
+
+	for (int channel_index = 0;
+			channel_index < config->num_channels;
+			channel_index++) {
+		flags = PIT_GetStatusFlags(config->base, channel_index);
+		if (flags) {
+			struct nxp_pit_channel_data *data =
+				PIT_CHANNEL_DATA(config->channels[channel_index]);
+			PIT_ClearStatusFlags(config->base, channel_index, flags);
+			data->top_callback(dev, data->top_user_data);
+		}
+	}
+}
+#else
+static void nxp_pit_isr(const struct device *dev)
+{
+	const struct nxp_pit_config *config = dev->config;
+	struct nxp_pit_channel_data *data = PIT_CHANNEL_DATA(dev);
+	pit_chnl_t channel = PIT_CHANNEL_ID(dev);
+	uint32_t flags;
+
+	LOG_DBG("pit counter isr");
+
+	flags = PIT_GetStatusFlags(config->base, channel);
+	if (flags) {
+		PIT_ClearStatusFlags(config->base, channel, flags);
+		data->top_callback(dev, data->top_user_data);
+	}
+}
+#endif /* DT_NODE_HAS_PROP(DT_COMPAT_GET_ANY_STATUS_OKAY(nxp_pit), interrupts) */
+
+static int nxp_pit_init(const struct device *dev)
+{
+	const struct nxp_pit_config *config = dev->config;
+	pit_config_t pit_config;
+	uint32_t clock_rate;
+
+	if (!device_is_ready(config->clock_dev)) {
+		LOG_ERR("Clock control device not ready");
+		return -ENODEV;
+	}
+
+	PIT_GetDefaultConfig(&pit_config);
+	pit_config.enableRunInDebug = config->enableRunInDebug;
+
+	PIT_Init(config->base, &pit_config);
+
+	clock_rate = nxp_pit_get_frequency(dev);
+
+#if DT_NODE_HAS_PROP(DT_COMPAT_GET_ANY_STATUS_OKAY(nxp_pit), interrupts)
+	config->irq_config_func(dev);
+	for (int channel_index = 0;
+		channel_index < config->num_channels;
+		channel_index++) {
+		PIT_SetTimerPeriod(config->base, channel_index,
+			USEC_TO_COUNT(config->info.max_top_value, clock_rate));
```

Comment:
nit: this code is the same between compilation paths, this source can be restructured to be simpler

Reply:
The code paths are not the same, in one code path a single interrupt config function is called, and in the other we ask the channel if they have an interrupt config function and if so we use their individual config functions, at the moment the driver will have either one case or the other.

Reply:
what I mean is something like this:
```
	for (int channel_index = 0;
		channel_index < config->num_channels;
		channel_index++) {
#if DT_NODE_HAS_PROP(DT_COMPAT_GET_ANY_STATUS_OKAY(nxp_pit), interrupts)
		if (config->irq_config_func[channel_index]) 
#endif
                {
			config->irq_config_func[channel_index](dev);
			PIT_SetTimerPeriod(config->base, channel_index,
				USEC_TO_COUNT(config->info.max_top_value, clock_rate));
		}
	}
```

but maybe check that compliance won't flag this

Reply:
okay, just forget about this suggestion, its really not important